graylog-storage-elasticsearch7/src/main/java/org/graylog/storage/elasticsearch7/IndicesAdapterES7.java

```diff
@@ -0,0 +1,474 @@
+                .subAggregation(AggregationBuilders.terms("streams").size(Integer.MAX_VALUE).field(Message.FIELD_STREAMS));
+        final SearchSourceBuilder query = SearchSourceBuilder.searchSource()
+                .aggregation(builder)
+                .size(0);
+
+        final SearchRequest request = new SearchRequest()
+                .source(query)
+                .indices(index)
+                .searchType(SearchType.DFS_QUERY_THEN_FETCH)
+                .indicesOptions(IndicesOptions.lenientExpandOpen());
+
+        final SearchResponse result = client.execute((c, requestOptions) -> c.search(request, requestOptions),
+                "Couldn't build index range of index " + index);
+
+        if (result.getTotalShards() == 0 || result.getAggregations() == null) {
+            throw new IndexNotFoundException("Couldn't build index range of index " + index + " because it doesn't exist.");
+        }
+        final Filter f = result.getAggregations().get("agg");
+        if (f == null) {
+            throw new IndexNotFoundException("Couldn't build index range of index " + index + " because it doesn't exist.");
+        } else if (f.getDocCount() == 0L) {
+            LOG.debug("No documents with attribute \"timestamp\" found in index <{}>", index);
+            return IndexRangeStats.EMPTY;
+        }
+
+        final Min minAgg = f.getAggregations().get("ts_min");
+        final long minUnixTime = new Double(minAgg.getValue()).longValue();
+        final DateTime min = new DateTime(minUnixTime, DateTimeZone.UTC);
+        final Max maxAgg = f.getAggregations().get("ts_max");
+        final long maxUnixTime = new Double(maxAgg.getValue()).longValue();
+        final DateTime max = new DateTime(maxUnixTime, DateTimeZone.UTC);
+        // make sure we return an empty list, so we can differentiate between old indices that don't have this information
+        // and newer ones that simply have no streams.
+        final Terms streams = f.getAggregations().get("streams");
+        final List<String> streamIds = streams.getBuckets().stream()
+                .map(MultiBucketsAggregation.Bucket::getKeyAsString)
+                .collect(toList());
+
+        return IndexRangeStats.create(min, max, streamIds);
+    }
+
+    @Override
+    public HealthStatus waitForRecovery(String index) {
+        final ClusterHealthRequest clusterHealthRequest = new ClusterHealthRequest(index);
+        clusterHealthRequest.waitForGreenStatus();
+
+        final ClusterHealthResponse result = client.execute((c, requestOptions) -> c.cluster().health(clusterHealthRequest, requestOptions));
+        return HealthStatus.fromString(result.getStatus().toString());
+    }
+
+    @Override
+    public boolean isOpen(String index) {
+        return indexState(index)
+                .map(state -> state.equals(State.Open))
+                .orElseThrow(() -> new IndexNotFoundException("Unable to determine state for absent index " + index));
```

Comment:
We should extract another private method here that gets `State` a parameter to remove the duplication with `isClosed`. Something like `indexHasState(String index, State state)`

Reply:
:heavy_check_mark: